From 01bf1005da937aad3666c17759d537637e3ad2d4 Mon Sep 17 00:00:00 2001
From: Luca Stefani <luca.stefani.ge1@gmail.com>
Date: Tue, 7 May 2019 19:55:34 +0200
Subject: [PATCH 6/6] [DNM] Squash of lineage-16.0-android-9.0.0_r37

Squashed commit of the following:

commit 94c1d740cfe001e5faf62d3c447c4e1e9f4aa325
Merge: 73ec8a7cc 1fb9beca3
Author: Bryan Ferris <bferris@google.com>
Date:   Fri Jan 4 15:36:52 2019 -0800

    [automerger skipped] DO NOT MERGE ANYWHERE resolve merge conflicts of bcf5e2388e608e4ec4372d1eeb918bbe0e095b2e to oc-dr1-dev am: 2ad06909c9 -s ours am skip reason: subject contains skip directive
    am: 1fb9beca3f -s ours
    am skip reason: subject contains skip directive

    Change-Id: Idf3e102c8176259eba32d253cd0d127f8d08d21b

commit 1fb9beca3f1d7ddc8a8add20dc84b43b262f5054
Merge: e730f5aaa 2ad06909c
Author: Bryan Ferris <bferris@google.com>
Date:   Fri Jan 4 15:33:51 2019 -0800

    [automerger skipped] DO NOT MERGE ANYWHERE resolve merge conflicts of bcf5e2388e608e4ec4372d1eeb918bbe0e095b2e to oc-dr1-dev
    am: 2ad06909c9 -s ours
    am skip reason: subject contains skip directive

    Change-Id: I08a81b0a363b7d31eb1bfbe1a64f24c2a78c8f67

commit 2ad06909c97d0ae17f2620bc6e2dbbbe36415bd5
Merge: bf8d7210c bcf5e2388
Author: Bryan Ferris <bferris@google.com>
Date:   Thu Jan 3 11:00:44 2019 -0800

    DO NOT MERGE ANYWHERE resolve merge conflicts of bcf5e2388e608e4ec4372d1eeb918bbe0e095b2e to oc-dr1-dev

    Bug: 115739809
    Bug: 120630087
    Test: I solemnly swear I tested this conflict resolution.
    Change-Id: If09327be8a4b4c35efe580d99d441d9320cbe2e9

commit 73ec8a7cc15268fd451b3b7cb6cf923b6741ce5f
Merge: ac1f7c9c5 e730f5aaa
Author: Siarhei Vishniakou <svv@google.com>
Date:   Wed Jan 2 17:41:43 2019 -0800

    Sanitize InputMessage before sending
    am: e730f5aaa1

    Change-Id: Iadf7bddb31cbef821417fc33b6c1d734b8dfc730

commit ac1f7c9c50b468315dbd4b0e71e03fb736fd98eb
Author: Siddharth Kapoor <ksiddharth@google.com>
Date:   Wed Dec 5 20:29:06 2018 +0800

    [RenderEngine] Clamp input color for BT2020_PQ EOTF

    Color input to EOTF for PQ transfer function is out-of-bounds [0.0, 1.0]
    Reason being, texture2D returns values which is not in range [0.0, 1.0].
    These returned values are passed as gl_FragColor.rgb to EOTF for PQ
    transfer function. In EOTF, colors outside the range [0.0, 1.0] are
    rendered as black resulting in artifacts while video playback.
    Clamp the input color to the range [0.0, 1.0]

    Bug: 118794307
    Test: Youtube HDR video plays fine, also screenshots taken while playback
      does not have any artifacts.

    Change-Id: I266f5a009c4c58924510283295757abae811a41e

commit e730f5aaa1c726ee9998a080e2d7f6284f4afec8
Author: Siarhei Vishniakou <svv@google.com>
Date:   Fri Nov 16 22:18:53 2018 -0800

    Sanitize InputMessage before sending

    The struct InputMessage has many fields, and is force-aligned to 8-byte
    boundaries. There are also some padding fields that carry no
    information.

    This struct is typically allocated in the stack and populated with
    various values before being sent across as a stream of bytes through the
    socket.

    Therefore, the "unused" data portions of the struct could contain
    portions of the stack, since there aren't ever writes to those memory
    locations.

    To avoid this information leak, forcefully sanitize the struct. Create a
    new struct that is explicitly set to zero. Next, only fill the
    meaningful fields manually.

    Bug: 115739809
    Test: cts-tradefed run cts -m CtsSecurityBulletinHostTestCases -t android.security.cts.Poc18_12; adb shell monkey 100000
    Change-Id: I7e44dacf1e8fa3156c8e4d2f7784ef0c53dab507
    Merged-In: I7e44dacf1e8fa3156c8e4d2f7784ef0c53dab507

commit bcf5e2388e608e4ec4372d1eeb918bbe0e095b2e
Merge: 4a1214cbb e67d59255
Author: Android Build Merger (Role) <noreply-android-build-merger@google.com>
Date:   Wed Dec 5 01:58:52 2018 +0000

    [automerger] Sanitize InputMessage before sending am: cb2f0ceed8 am: 9a7bf9f1db am: 0e9e9b0970 am: 26bf1d2499 am: e67d59255e

    Change-Id: Icb796e5fec430490c211ecfc365f38c0897c5d79

commit e67d59255e5aab7ca1ea4a2893c2ab88d0b70d15
Merge: 5b0043ff0 26bf1d249
Author: Android Build Merger (Role) <noreply-android-build-merger@google.com>
Date:   Wed Dec 5 01:58:49 2018 +0000

    [automerger] Sanitize InputMessage before sending am: cb2f0ceed8 am: 9a7bf9f1db am: 0e9e9b0970 am: 26bf1d2499

    Change-Id: I37de2ace6d13d022d63c11446a4a58ec929efa6f

commit 26bf1d249948488e01a1f8afcbd773989b7f5567
Merge: e4971bf98 0e9e9b097
Author: Android Build Merger (Role) <noreply-android-build-merger@google.com>
Date:   Wed Dec 5 01:58:47 2018 +0000

    [automerger] Sanitize InputMessage before sending am: cb2f0ceed8 am: 9a7bf9f1db am: 0e9e9b0970

    Change-Id: I4fcda65956abf2331e0625ec71b76e32322584ef

commit 0e9e9b09701897959e004c689574e173cf70dab4
Merge: d3159fdd1 9a7bf9f1d
Author: Android Build Merger (Role) <noreply-android-build-merger@google.com>
Date:   Wed Dec 5 01:58:44 2018 +0000

    [automerger] Sanitize InputMessage before sending am: cb2f0ceed8 am: 9a7bf9f1db

    Change-Id: I4f797beb8c68c7442a908313ffb9d71c42ef5eec

commit 9a7bf9f1dbe6a28d5f6188388c2be07f0834cf5d
Merge: 42eda1ed4 cb2f0ceed
Author: Android Build Merger (Role) <noreply-android-build-merger@google.com>
Date:   Wed Dec 5 01:58:42 2018 +0000

    [automerger] Sanitize InputMessage before sending am: cb2f0ceed8

    Change-Id: Id5fe663812bb4e355656cb0715ce4e56802e32e8

commit cb2f0ceed876b3c1cc62fc677c75a12c1f10b199
Author: Siarhei Vishniakou <svv@google.com>
Date:   Fri Nov 16 22:18:53 2018 -0800

    Sanitize InputMessage before sending

    The struct InputMessage has many fields, and is force-aligned to 8-byte
    boundaries. There are also some padding fields that carry no
    information.

    This struct is typically allocated in the stack and populated with
    various values before being sent across as a stream of bytes through the
    socket.

    Therefore, the "unused" data portions of the struct could contain
    portions of the stack, since there aren't ever writes to those memory
    locations.

    To avoid this information leak, forcefully sanitize the struct. Create a
    new struct that is explicitly set to zero. Next, only fill the
    meaningful fields manually.

    Bug: 115739809
    Test: cts-tradefed run cts -m CtsSecurityBulletinHostTestCases -t android.security.cts.Poc18_12; adb shell monkey 100000
    Change-Id: I7e44dacf1e8fa3156c8e4d2f7784ef0c53dab507
    Merged-In: I7e44dacf1e8fa3156c8e4d2f7784ef0c53dab507

commit 5b0043ff0cc1f71ddec8f39efbfa570387a3529f
Merge: 634696f9e 04ab0619b
Author: TreeHugger Robot <treehugger-gerrit@google.com>
Date:   Fri May 11 09:53:53 2018 +0000

    Merge changes from topic "am-27943b02-3710-49e4-84f6-eabc78fdbe01" into nyc-mr2-dev

    * changes:
      [automerger] Don't pad before calling writeInPlace(). am: 732132b765 am: a6a5c3fa47 am: 325909a4e4 am: 953b08b7c0 am: 32da4bbe1e am: a01397282d am: 6fccb21cd7 am: f6384b4a10
      [automerger] Don't pad before calling writeInPlace(). am: 732132b765 am: a6a5c3fa47 am: 325909a4e4 am: 953b08b7c0 am: 32da4bbe1e am: a01397282d am: 6fccb21cd7
      [automerger] Don't pad before calling writeInPlace(). am: 732132b765 am: a6a5c3fa47 am: 325909a4e4 am: 953b08b7c0 am: 32da4bbe1e am: a01397282d
      [automerger] Don't pad before calling writeInPlace(). am: 732132b765 am: a6a5c3fa47 am: 325909a4e4 am: 953b08b7c0 am: 32da4bbe1e
      [automerger] Don't pad before calling writeInPlace(). am: 732132b765 am: a6a5c3fa47 am: 325909a4e4 am: 953b08b7c0
      [automerger] Don't pad before calling writeInPlace(). am: 732132b765 am: a6a5c3fa47 am: 325909a4e4
      [automerger] Don't pad before calling writeInPlace(). am: 732132b765 am: a6a5c3fa47
      [automerger] Don't pad before calling writeInPlace(). am: 732132b765
      Don't pad before calling writeInPlace().

commit e4971bf98e0f5f06a4ab73bd2b7a07c7e1fbd16b
Merge: e1794e5b4 f6384b4a1
Author: TreeHugger Robot <treehugger-gerrit@google.com>
Date:   Fri May 11 09:53:53 2018 +0000

    Merge changes from topic "am-27943b02-3710-49e4-84f6-eabc78fdbe01" into cw-f-dev

    * changes:
      [automerger] Don't pad before calling writeInPlace(). am: 732132b765 am: a6a5c3fa47 am: 325909a4e4 am: 953b08b7c0 am: 32da4bbe1e am: a01397282d am: 6fccb21cd7
      [automerger] Don't pad before calling writeInPlace(). am: 732132b765 am: a6a5c3fa47 am: 325909a4e4 am: 953b08b7c0 am: 32da4bbe1e am: a01397282d
      [automerger] Don't pad before calling writeInPlace(). am: 732132b765 am: a6a5c3fa47 am: 325909a4e4 am: 953b08b7c0 am: 32da4bbe1e
      [automerger] Don't pad before calling writeInPlace(). am: 732132b765 am: a6a5c3fa47 am: 325909a4e4 am: 953b08b7c0
      [automerger] Don't pad before calling writeInPlace(). am: 732132b765 am: a6a5c3fa47 am: 325909a4e4
      [automerger] Don't pad before calling writeInPlace(). am: 732132b765 am: a6a5c3fa47
      [automerger] Don't pad before calling writeInPlace(). am: 732132b765
      Don't pad before calling writeInPlace().

commit d3159fdd10f9f025e9232798cebdcdc064c343fc
Merge: d9d1fbf29 6fccb21cd
Author: TreeHugger Robot <treehugger-gerrit@google.com>
Date:   Fri May 11 09:53:53 2018 +0000

    Merge changes from topic "am-27943b02-3710-49e4-84f6-eabc78fdbe01" into nyc-mr1-dev

    * changes:
      [automerger] Don't pad before calling writeInPlace(). am: 732132b765 am: a6a5c3fa47 am: 325909a4e4 am: 953b08b7c0 am: 32da4bbe1e am: a01397282d
      [automerger] Don't pad before calling writeInPlace(). am: 732132b765 am: a6a5c3fa47 am: 325909a4e4 am: 953b08b7c0 am: 32da4bbe1e
      [automerger] Don't pad before calling writeInPlace(). am: 732132b765 am: a6a5c3fa47 am: 325909a4e4 am: 953b08b7c0
      [automerger] Don't pad before calling writeInPlace(). am: 732132b765 am: a6a5c3fa47 am: 325909a4e4
      [automerger] Don't pad before calling writeInPlace(). am: 732132b765 am: a6a5c3fa47
      [automerger] Don't pad before calling writeInPlace(). am: 732132b765
      Don't pad before calling writeInPlace().

commit 42eda1ed42a7f1cc751dd57b858be75b06179dae
Merge: 75061f171 a01397282
Author: TreeHugger Robot <treehugger-gerrit@google.com>
Date:   Fri May 11 09:53:53 2018 +0000

    Merge changes from topic "am-27943b02-3710-49e4-84f6-eabc78fdbe01" into nyc-dr1-dev

    * changes:
      [automerger] Don't pad before calling writeInPlace(). am: 732132b765 am: a6a5c3fa47 am: 325909a4e4 am: 953b08b7c0 am: 32da4bbe1e
      [automerger] Don't pad before calling writeInPlace(). am: 732132b765 am: a6a5c3fa47 am: 325909a4e4 am: 953b08b7c0
      [automerger] Don't pad before calling writeInPlace(). am: 732132b765 am: a6a5c3fa47 am: 325909a4e4
      [automerger] Don't pad before calling writeInPlace(). am: 732132b765 am: a6a5c3fa47
      [automerger] Don't pad before calling writeInPlace(). am: 732132b765
      Don't pad before calling writeInPlace().

commit f7159a92e2bcfb02dd24469a75def62e7c6e35f1
Merge: 4af41a297 32da4bbe1
Author: TreeHugger Robot <treehugger-gerrit@google.com>
Date:   Fri May 11 09:53:53 2018 +0000

    Merge changes from topic "am-27943b02-3710-49e4-84f6-eabc78fdbe01" into nyc-dev

    * changes:
      [automerger] Don't pad before calling writeInPlace(). am: 732132b765 am: a6a5c3fa47 am: 325909a4e4 am: 953b08b7c0
      [automerger] Don't pad before calling writeInPlace(). am: 732132b765 am: a6a5c3fa47 am: 325909a4e4
      [automerger] Don't pad before calling writeInPlace(). am: 732132b765 am: a6a5c3fa47
      [automerger] Don't pad before calling writeInPlace(). am: 732132b765
      Don't pad before calling writeInPlace().

Change-Id: Ie8cd64b1bcedb5f3f6eba0864ed90789937c4c9c
---
 services/surfaceflinger/RenderEngine/ProgramCache.cpp | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/services/surfaceflinger/RenderEngine/ProgramCache.cpp b/services/surfaceflinger/RenderEngine/ProgramCache.cpp
index 9dc685856..46402d516 100644
--- a/services/surfaceflinger/RenderEngine/ProgramCache.cpp
+++ b/services/surfaceflinger/RenderEngine/ProgramCache.cpp
@@ -220,7 +220,7 @@ void ProgramCache::generateEOTF(Formatter& fs, const Key& needs) {
                     const highp float c2 = (2413.0 / 4096.0) * 32.0;
                     const highp float c3 = (2392.0 / 4096.0) * 32.0;
 
-                    highp vec3 tmp = pow(color, 1.0 / vec3(m2));
+                    highp vec3 tmp = pow(clamp(color, 0.0, 1.0), 1.0 / vec3(m2));
                     tmp = max(tmp - c1, 0.0) / (c2 - c3 * tmp);
                     return pow(tmp, 1.0 / vec3(m1));
                 }
-- 
2.17.1

